### PR TITLE
adding metrics system tests

### PR DESCRIPTION
**Pull request by franvila**
### Type of change

- Enhancement / new feature

### Description

Adding metrics system tests to cover #918 

### Additional Context

Adapted what strimzi system tests already has to cover what is needed for Kroxylicious.

### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [X] Write tests
- [ ] Make sure all tests pass
- [ ] Review performance test results. Ensure that any degradations to performance numbers are understood and justified.
- [ ] Make sure all Sonarcloud warnings are addressed or are justifiably ignored.
- [ ] Update documentation
- [ ] Reference relevant issue(s) and close them after merging
- [ ] For user facing changes, update CHANGELOG.md (remember to include changes affecting the API of the test artefacts too).

### CODE REVIEW

File: kroxylicious-systemtests/src/main/java/io/kroxylicious/systemtests/Constants.java [hunk 1]
```diff
@@ -112,11 +112,16 @@ public interface Constants {
     /**
      * Global timeout in milliseconds
      */
-    long GLOBAL_TIMEOUT_MILLIS = Duration.ofMinutes(5).toMillis();
+    long GLOBAL_TIMEOUT = Duration.ofMinutes(5).toMillis();
```
Comment (k-wall) [inline kroxylicious-systemtests/src/main/java/io/kroxylicious/systemtests/Constants.java:115]: minor: the old name was better imho.   Alternative, keep them as Duration objects here, and that way there's no need to encode the unit in the variable name.

File: kroxylicious-systemtests/src/main/java/io/kroxylicious/systemtests/metrics/MetricsCollector.java [hunk 2]
```diff
@@ -0,0 +1,387 @@
...
+    private final String componentName;
+    private final int metricsPort;
+    private final String metricsPath;
+    private final LabelSelector componentLabelSelector;
+    private Map<String, String> collectedData;
+
+    /**
+     * The type Builder.
+     */
+    public static class Builder {
```
Comment (k-wall) [inline kroxylicious-systemtests/src/main/java/io/kroxylicious/systemtests/metrics/MetricsCollector.java:53]: does the build need to be public?

File: kroxylicious-systemtests/src/main/java/io/kroxylicious/systemtests/metrics/MetricsCollector.java [hunk 3]
```diff
@@ -0,0 +1,387 @@
...
+        namespaceName = builder.namespaceName;
+        scraperPodName = builder.scraperPodName;
+        metricsPort = builder.metricsPort;
+        metricsPath = builder.metricsPath;
+        componentName = builder.componentName;
+        componentLabelSelector = getLabelSelectorForResource();
+    }
+
+    private LabelSelector getLabelSelectorForResource() {
+        if (Objects.requireNonNull(this.componentType) == ComponentType.Kroxylicious) {
```
Comment (k-wall) [inline kroxylicious-systemtests/src/main/java/io/kroxylicious/systemtests/metrics/MetricsCollector.java:261]: ouch! that violates least surprise. by about a parsec. 

etters shouldn't make assertions about state, even if they are private.  Best to impose the condition on construction, so it is easy for the caller to know where they went wrong.

File: kroxylicious-systemtests/src/main/java/io/kroxylicious/systemtests/metrics/MetricsCollector.java [hunk 4]
```diff
@@ -0,0 +1,387 @@
...
+    /**
+     * Method checks already collected metrics data for Pattern containing desired metric
+     * @param pattern Pattern of metric which is desired
+     * @return ArrayList of values collected from the metrics
+     */
+    public synchronized ArrayList<Double> waitForSpecificMetricAndCollect(Pattern pattern) {
+        ArrayList<Double> values = collectSpecificMetric(pattern);
+
+        if (values.isEmpty()) {
+            TestUtils.waitFor(String.format("metrics contain pattern: %s", pattern.toString()), Constants.GLOBAL_POLL_INTERVAL_MEDIUM, Constants.GLOBAL_STATUS_TIMEOUT,
```
Comment (k-wall) [inline kroxylicious-systemtests/src/main/java/io/kroxylicious/systemtests/metrics/MetricsCollector.java:304]: we should probably prefer Awaitility

File: kroxylicious-systemtests/src/main/java/io/kroxylicious/systemtests/metrics/MetricsCollector.java [hunk 5]
```diff
@@ -0,0 +1,387 @@
...
+    private String collectMetrics(String metricsPodIp, String podName) throws InterruptedException, ExecutionException, IOException {
+        List<String> executableCommand = Arrays.asList(cmdKubeClient(namespaceName).toString(), "exec", scraperPodName,
+                "-n", namespaceName,
+                "--", "curl", metricsPodIp + ":" + metricsPort + metricsPath);
+
+        LOGGER.debug("Executing command:{} for scrape the metrics", executableCommand);
+
+        Exec exec = new Exec();
+        // 20 seconds should be enough for collect data from the pod
+        int ret = exec.execute(null, executableCommand, 20_000, null);
```
Comment (k-wall) [inline kroxylicious-systemtests/src/main/java/io/kroxylicious/systemtests/metrics/MetricsCollector.java:335]: probably want to check this is zero.

File: kroxylicious-systemtests/src/main/java/io/kroxylicious/systemtests/utils/MetricsUtils.java [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+ *  Provides auxiliary methods for Scraper Pod, which reaches KafkaConnect API in the Kubernetes cluster.
+ */
+public class MetricsUtils {
+    private static final Logger LOGGER = LogManager.getLogger(MetricsUtils.class);
+    private static final Object LOCK = new Object();
+
+    private MetricsUtils() {
+    }
+
+    public static String getExporterRunScript(String podName, String namespace) throws InterruptedException, ExecutionException, IOException {
```
Comment (k-wall) [inline kroxylicious-systemtests/src/main/java/io/kroxylicious/systemtests/utils/MetricsUtils.java:40]: Do we need this?

File: kroxylicious-systemtests/src/test/java/io/kroxylicious/systemtests/MetricsST.java [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+ */
+@ExtendWith(KroxyliciousExtension.class)
+public class MetricsST extends AbstractST {
+    private static final Logger LOGGER = LogManager.getLogger(MetricsST.class);
+    private final String clusterName = "my-cluster";
+    protected static final String BROKER_NODE_NAME = "kafka";
+    private static final String MESSAGE = "Hello-world";
+    private static Kroxylicious kroxylicious;
+    private MetricsCollector kroxyliciousCollector;
+
```
Comment (k-wall) [inline kroxylicious-systemtests/src/test/java/io/kroxylicious/systemtests/MetricsST.java:49]: You've tested just a couple of the metrics exposed by Kroxy.  I think this is good enough for now.

File: kroxylicious-systemtests/src/main/java/io/kroxylicious/systemtests/Constants.java [hunk 8]
```diff
@@ -156,4 +161,12 @@ public interface Constants {
...
     String USE_KAFKA_NODE_POOLS = "+KafkaNodePools";
     String DONT_USE_KAFKA_NODE_POOLS = "-KafkaNodePools";
+
+    /**
+     * Scraper pod labels
+     */
+    String SCRAPER_LABEL_KEY = "user-test-app";
+    String SCRAPER_LABEL_VALUE = "scraper";
+    String SCRAPER_NAME = "scraper";
+    String DEPLOYMENT_TYPE = "deployment-type";
```
Comment (robobario) [inline kroxylicious-systemtests/src/main/java/io/kroxylicious/systemtests/Constants.java:171]:
```suggestion
    String DEPLOYMENT_TYPE_LABEL_KEY = "deployment-type";
```
and could make the `Scraper` deployment type a constant (or use SCRAPER_NAME)

File: kroxylicious-systemtests/src/main/java/io/kroxylicious/systemtests/k8s/KubeClusterResource.java [hunk 9]
```diff
@@ -68,6 +68,18 @@ public String getNamespace() {
...
     }
 
+    /**
+     * Sets the namespace value for Kubernetes clients
+     * @param futureNamespace Namespace which should be used in Kubernetes clients
+     * @return Previous namespace which was used in Kubernetes clients
+     */
+    public String setNamespace(String futureNamespace) {
+        String previousNamespace = namespace;
+        LOGGER.info("Client use Namespace: {}", futureNamespace);
```
Comment (robobario) [inline kroxylicious-systemtests/src/main/java/io/kroxylicious/systemtests/k8s/KubeClusterResource.java:78]: should we log the previousNamespace while we're here?

File: kroxylicious-systemtests/src/main/java/io/kroxylicious/systemtests/Constants.java (hunk 1 above)
Comment (franvila) [reply kroxylicious-systemtests/src/main/java/io/kroxylicious/systemtests/Constants.java]: Transformed all timeouts to `Duration` type

File: kroxylicious-systemtests/src/test/java/io/kroxylicious/systemtests/MetricsST.java (hunk 7 above)
Comment (franvila) [reply kroxylicious-systemtests/src/test/java/io/kroxylicious/systemtests/MetricsST.java:49]: Me too

File: kroxylicious-systemtests/src/main/java/io/kroxylicious/systemtests/metrics/MetricsCollector.java (hunk 2 above)
Comment (franvila) [reply kroxylicious-systemtests/src/main/java/io/kroxylicious/systemtests/metrics/MetricsCollector.java:54]: The way it is used in the beforeEach in the MetricsST class is needed to be public. I didn't waste time to review all what is done in Strimzi, I have just copied and adapted as much as possible.

File: kroxylicious-systemtests/src/main/java/io/kroxylicious/systemtests/metrics/MetricsCollector.java [hunk 10]
```diff
@@ -235,31 +236,24 @@ public MetricsCollector.Builder toBuilder() {
...
-            builder.metricsPath = getDefaultMetricsPath();
+        if (Optional.ofNullable(builder.scraperPodName).isEmpty()) {
+            throw new InvalidParameterException("Scraper Pod name is not set");
+        }
 
-        namespaceName = builder.namespaceName;
         scraperPodName = builder.scraperPodName;
-        metricsPort = builder.metricsPort;
-        metricsPath = builder.metricsPath;
+        namespaceName = (Optional.ofNullable(builder.namespaceName).isEmpty()) ? kubeClient().getNamespace() : builder.namespaceName;
```
Comment (k-wall) [inline kroxylicious-systemtests/src/main/java/io/kroxylicious/systemtests/metrics/MetricsCollector.java:246]:
```suggestion
        namespaceName = Optional.ofNullable(builder.namespaceName).orElse(kubeClient().getNamespace());
```

Comment (franvila) [reply kroxylicious-systemtests/src/main/java/io/kroxylicious/systemtests/metrics/MetricsCollector.java:246]: `orElse` is also checking if the string is empty? or only if it is null?

File: kroxylicious-systemtests/src/main/java/io/kroxylicious/systemtests/metrics/MetricsCollector.java [hunk 11]
```diff
@@ -345,8 +339,8 @@ private String collectMetrics(String metricsPodIp, String podName) throws Interr
     @SuppressWarnings("unchecked")
     public void collectMetricsFromPods() {
         Map<String, String>[] metricsData = (Map<String, String>[]) new HashMap[1];
-        TestUtils.waitFor("metrics to contain data", Constants.GLOBAL_POLL_INTERVAL, Constants.GLOBAL_TIMEOUT,
-                () -> {
+        await().atMost(Constants.GLOBAL_TIMEOUT).pollInterval(Constants.GLOBAL_POLL_INTERVAL)
+                .until(() -> {
```
Comment (k-wall) [inline kroxylicious-systemtests/src/main/java/io/kroxylicious/systemtests/metrics/MetricsCollector.java:343]: use the two arg form of until so you can access what was produced.  that gets rid of the that odd array of hashmaps.

(untested code)
```
        collectedData = await().atMost(Constants.GLOBAL_TIMEOUT).pollInterval(Constants.GLOBAL_POLL_INTERVAL)
                .until(this::collectMetricsFromPodsWithoutWait,
                        (collected) -> {
                            // KafkaExporter metrics should be non-empty
                            if (collectedData.isEmpty()) {
                                return false;
                            }

                            for (Map.Entry<String, String> item : collected.entrySet()) {
                                if (item.getValue().isEmpty()) {
                                    return false;
                                }
                            }
                            return true;
                        });
```

Comment (k-wall) [reply kroxylicious-systemtests/src/main/java/io/kroxylicious/systemtests/metrics/MetricsCollector.java:343]: you should be able to reduce the second block to something like:

```
   return !(collected.isEmpty() || collected.entrySet().stream().anyMatch(e -> e.getValue().isEmpty()));
```

File: kroxylicious-systemtests/src/main/java/io/kroxylicious/systemtests/metrics/MetricsCollector.java [hunk 12]
```diff
@@ -330,6 +325,11 @@ private String collectMetrics(String metricsPodIp, String podName) throws Interr
 
         LOGGER.info("Metrics collection for Pod: {}/{}({}) from Pod: {}/{} finished with return code: {}", namespaceName, podName, metricsPodIp, namespaceName,
                 scraperPodName, ret);
+
+        if (ret != 0) {
+            throw new IOException("Cannot collect the metrics using " + executableCommand + " command");
```
Comment (k-wall) [inline kroxylicious-systemtests/src/main/java/io/kroxylicious/systemtests/metrics/MetricsCollector.java:330]: it is not a IOException, so don't suggest it is 

```suggestion
            throw new RuntimeException("Cannot collect the metrics using " + executableCommand + " command");
```

**Comment by k-wall** [comment]
@franvila thank you for the work.  It is worth trying address the quality gate issues, then I'd be happy to +1

File: kroxylicious-systemtests/src/main/java/io/kroxylicious/systemtests/executor/Exec.java [hunk 13]
```diff
@@ -293,14 +294,14 @@ private static void logExecutor(int ret, String input, List<String> command, Str
      *
      * @param input the input
      * @param commands arguments for command
-     * @param timeoutMs timeout in ms for kill
+     * @param timeout timeout in for kill
```
Comment (k-wall) [inline kroxylicious-systemtests/src/main/java/io/kroxylicious/systemtests/executor/Exec.java:297]:
```suggestion
     * @param timeout timeout allowed for the process's execution. If the timeout is exceed the process will be killed.
```

File: kroxylicious-systemtests/src/test/java/io/kroxylicious/systemtests/MetricsST.java [hunk 14]
```diff
@@ -0,0 +1,141 @@
...
+        Kafka kafka = KafkaTemplates.kafkaPersistentWithKRaftAnnotations(Constants.KAFKA_DEFAULT_NAMESPACE, clusterName, 3).build();
+
+        resourceManager.createResourceWithWait(
+                KafkaNodePoolTemplates.kafkaBasedNodePoolWithDualRole(BROKER_NODE_NAME, kafka, 3).build(),
+                kafka);
+    }
+
+    @BeforeEach
+    void beforeEach(String namespace) throws InterruptedException {
```
Comment (k-wall) [inline kroxylicious-systemtests/src/test/java/io/kroxylicious/systemtests/MetricsST.java:121]:
```suggestion
    @SuppressWarnings("java:S2925")
    void beforeEach(String namespace) throws InterruptedException {
```

File: kroxylicious-systemtests/src/main/java/io/kroxylicious/systemtests/resources/ComponentType.java [hunk 15]
```diff
@@ -0,0 +1,11 @@
+/*
+ * Copyright Kroxylicious Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+package io.kroxylicious.systemtests.resources;
+
+public enum ComponentType {
+    Kroxylicious
```
Comment (k-wall) [inline kroxylicious-systemtests/src/main/java/io/kroxylicious/systemtests/resources/ComponentType.java:10]: That's breaking the naming convention - make it all capitals.

Comment (franvila) [reply kroxylicious-systemtests/src/main/java/io/kroxylicious/systemtests/resources/ComponentType.java]: I was working on latest code smells

**Comment by k-wall** [comment]
I've used overridden the rule to allow for our system test naming convention. https://sonarcloud.io/organizations/kroxylicious/rules?activation=true&inheritance=OVERRIDES&open=java%3AS3577&qprofile=AYtEnAUgVN5wppgWgCsF

File: kroxylicious-systemtests/src/test/java/io/kroxylicious/systemtests/MetricsST.java [hunk 16]
```diff
@@ -0,0 +1,142 @@
...
+import io.kroxylicious.systemtests.templates.strimzi.KafkaTemplates;
+
+import static io.kroxylicious.systemtests.k8s.KubeClusterResource.kubeClient;
+import static io.kroxylicious.systemtests.utils.MetricsUtils.assertMetricValue;
+import static io.kroxylicious.systemtests.utils.MetricsUtils.assertMetricValueCount;
+import static io.kroxylicious.systemtests.utils.MetricsUtils.assertMetricValueHigherThan;
+import static org.junit.jupiter.api.Assertions.assertAll;
+
+/**
+ * @description This test suite is designed for testing metrics exposed by kroxylicious.
```
Comment (k-wall) [inline kroxylicious-systemtests/src/test/java/io/kroxylicious/systemtests/MetricsST.java:39]: `@description` isn't a recognised javadoc tag.
```suggestion
 * This test suite is designed for testing metrics exposed by kroxylicious.
```

File: kroxylicious-systemtests/src/test/java/io/kroxylicious/systemtests/MetricsST.java [hunk 17]
```diff
@@ -0,0 +1,142 @@
...
+class MetricsST extends AbstractST {
+    private static final Logger LOGGER = LogManager.getLogger(MetricsST.class);
+    private final String clusterName = "my-cluster";
+    protected static final String BROKER_NODE_NAME = "kafka";
+    private static final String MESSAGE = "Hello-world";
+    private static Kroxylicious kroxylicious;
+    private MetricsCollector kroxyliciousCollector;
+
+    @Test
+    void testKroxyliciousMetricsBeforeSendingMessages() {
```
Comment (k-wall) [inline kroxylicious-systemtests/src/test/java/io/kroxylicious/systemtests/MetricsST.java:51]:
```suggestion
    void kroxyliciousMetricsBeforeSendingMessages() {
```

File: kroxylicious-systemtests/src/test/java/io/kroxylicious/systemtests/MetricsST.java [hunk 18]
```diff
@@ -0,0 +1,142 @@
...
+        assertAll("Checking the presence of the metrics",
+                () -> assertMetricValueCount(kroxyliciousCollector, "kroxylicious_inbound_downstream_messages_total", 1),
+                () -> assertMetricValueCount(kroxyliciousCollector, "kroxylicious_inbound_downstream_decoded_messages_total", 1));
+        assertAll("Checking the value of the metrics",
+                () -> assertMetricValue(kroxyliciousCollector, "kroxylicious_inbound_downstream_messages_total", 0),
+                () -> assertMetricValue(kroxyliciousCollector, "kroxylicious_inbound_downstream_decoded_messages_total", 0));
+    }
+
+    @Test
+    void testKroxyliciousDownstreamMessages(String namespace) {
```
Comment (k-wall) [inline kroxylicious-systemtests/src/test/java/io/kroxylicious/systemtests/MetricsST.java:62]:
```suggestion
    void kroxyliciousDownstreamMessages(String namespace) {
```

File: kroxylicious-systemtests/src/test/java/io/kroxylicious/systemtests/MetricsST.java [hunk 19]
```diff
@@ -0,0 +1,142 @@
...
+        kroxyliciousCollector.collectMetricsFromPods();
+        LOGGER.info("Metrics: " + kroxyliciousCollector.getCollectedData().values());
+        assertAll(
+                () -> assertMetricValueCount(kroxyliciousCollector, "kroxylicious_inbound_downstream_messages_total", 1),
+                () -> assertMetricValueHigherThan(kroxyliciousCollector, "kroxylicious_inbound_downstream_messages_total", 0),
+                () -> assertMetricValueHigherThan(kroxyliciousCollector, "kroxylicious_inbound_downstream_decoded_messages_total", 0));
+    }
+
+    @Test
+    void testKroxyliciousPayloadSize(String namespace) {
```
Comment (k-wall) [inline kroxylicious-systemtests/src/test/java/io/kroxylicious/systemtests/MetricsST.java:83]:
```suggestion
    void kroxyliciousPayloadSize(String namespace) {
```

**Review by k-wall — APPROVED** [review]
Ran tests locally, they are passing okay for me.
This looks good enough for now.
Thank you @franvila

**Comment by franvila** [comment]
@strimzi-ci run system test

**Comment by franvila** [comment]
@strimzi-ci run system tests

**Comment by franvila** [comment]
@strimzi-ci run system tests

**Review by robobario — APPROVED** [review]
LGTM

**Pull request merged**